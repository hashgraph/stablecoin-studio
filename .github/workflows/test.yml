name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test