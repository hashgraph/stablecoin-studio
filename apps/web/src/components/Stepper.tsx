import {
	Tabs as ChakraTabs,
	TabList,
	TabPanels,
	Tab,
	TabPanel,
	Text,
	Box,
	Divider,
	Flex,
	Button,
	VStack,
} from '@chakra-ui/react';
import type { MouseEvent, ReactNode } from 'react';
import { useTranslation } from 'react-i18next';

export interface Step {
	number: string;
	title: string;
	children: ReactNode;
}

interface StepperProps {
	isValid: boolean;
	steps: Step[];
	textDefaultButtonPrimary?: string;
	textLastButtonPrimary?: string;
	textDefaultButtonSecondary?: string;
	textFirstButtonSecondary?: string;
	currentStep: number;
	setCurrentStep: (arg0: number) => void;
	handleFirstButtonSecondary: () => void;
	handleLastButtonPrimary: () => void;
	supplyKey?: number;
}

const FOOTER_HEIGHT = '88px';

const Stepper = (props: StepperProps) => {
	const { t } = useTranslation('global');

	const {
		isValid,
		steps,
		textDefaultButtonPrimary = t('stepper.nextStep'),
		textLastButtonPrimary = t('stepper.finish'),
		textDefaultButtonSecondary = t('stepper.goBack'),
		textFirstButtonSecondary = t('common.cancel'),
		currentStep,
		setCurrentStep,
		handleFirstButtonSecondary,
		handleLastButtonPrimary,
		supplyKey,
	} = props;

	const handleStep = (e: MouseEvent<HTMLButtonElement>, index: number, type: 'next' | 'prev') => {
		e.preventDefault();

		if (supplyKey !== 2) {
			return type === 'next'
				? index === 2
					? setCurrentStep(4)
					: setCurrentStep(index + 1)
				: index === 4
				? setCurrentStep(2)
				: setCurrentStep(index - 1);
		} else {
			return type === 'next' ? setCurrentStep(index + 1) : setCurrentStep(index - 1);
		}
	};

	return (
		<ChakraTabs isFitted isLazy variant='simple' index={currentStep} position='relative' h='full'>
			<TabList justifyContent='space-between'>
				{steps.map((step, index) => {
					const { title, number } = step;

					const isCurrentStep = currentStep === index;

					const handleChangeTab = () => {
						setCurrentStep(index);
					};
					return (
						<Tab
							key={index}
							p='15px'
							isDisabled={(supplyKey !== 2 && index === 3) || index > currentStep}
							onClick={handleChangeTab}
							_hover={{
								cursor: index < currentStep ? 'pointer' : 'default',
							}}
						>
							<Text
								data-testid={`stepper-step-number-${index + 1}`}
								fontSize='14px'
								color={isCurrentStep ? 'brand.primary' : 'brand.gray600'}
								mr='10px'
							>
								{number}
							</Text>
							<Text
								data-testid={`stepper-step-title-${index + 1}`}
								fontSize='14px'
								color='brand.gray600'
							>
								{title}
							</Text>
						</Tab>
					);
				})}
			</TabList>

			<TabPanels h='calc(100% - 55px)' bgColor='brand.gray100'>
				{steps.map((step, index) => {
					const { children } = step;

					const isFirstStep = index === 0;
					const isLastStep = index === steps.length - 1;

					return (
						<TabPanel
							data-testid={`stepper-step-panel-container-${index + 1}`}
							h='full'
							p='0px'
							key={index}
						>
							<VStack h='full' spacing={4} align='stretch'>
								<Flex flex={1} justify='center' overflowY='auto'>
									{children}
								</Flex>
								<Box h={FOOTER_HEIGHT}>
									<Divider color='brand.gray2' />
									<Flex p='0px 24px' justifyContent='flex-end' h='99%' alignItems='center'>
										<Button
											data-testid={`stepper-step-panel-button-secondary-${index + 1}`}
											variant='secondary'
											mr='20px'
											onClick={
												isFirstStep
													? handleFirstButtonSecondary
													: (e) => handleStep(e, index, 'prev')
											}
										>
											{isFirstStep ? textFirstButtonSecondary : textDefaultButtonSecondary}
										</Button>
										<Button
											data-testid={`stepper-step-panel-button-primary-${index + 1}`}
											isDisabled={!isValid}
											variant='primary'
											onClick={
												isLastStep ? handleLastButtonPrimary : (e) => handleStep(e, index, 'next')
											}
										>
											{isLastStep ? textLastButtonPrimary : textDefaultButtonPrimary}
										</Button>
									</Flex>
								</Box>
							</VStack>
						</TabPanel>
					);
				})}
			</TabPanels>
		</ChakraTabs>
	);
};

export default Stepper;
