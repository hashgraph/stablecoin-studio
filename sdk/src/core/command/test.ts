/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
