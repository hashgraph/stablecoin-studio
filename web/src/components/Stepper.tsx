import {
	Tabs as ChakraTabs,
	TabList,
	TabPanels,
	Tab,
	TabPanel,
	Text,
	Box,
	Divider,
	Flex,
	Button,
	VStack,
} from '@chakra-ui/react';
import type { MouseEvent, ReactNode } from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';

export interface Step {
	number: string;
	title: string;
	complete: boolean;
	children: ReactNode;
}

interface StepperProps {
	isValid: boolean;
	steps: Step[];
	textDefaultButtonPrimary?: string;
	textLastButtonPrimary?: string;
	textDefaultButtonSecondary?: string;
	textFirstButtonSecondary?: string;
	handleFirstButtonSecondary: () => void;
	handleLastButtonPrimary: () => void;
}

const FOOTER_HEIGHT = '88px';

const Stepper = (props: StepperProps) => {
	const { t } = useTranslation('global');

	const {
		isValid,
		steps,
		textDefaultButtonPrimary = t('stepper.nextStep'),
		textLastButtonPrimary = t('stepper.finish'),
		textDefaultButtonSecondary = t('stepper.goBack'),
		textFirstButtonSecondary = t('common.cancel'),
		handleFirstButtonSecondary,
		handleLastButtonPrimary,
	} = props;

	const [currentStep, setCurrentStep] = useState(0);

	const handleStep = (e: MouseEvent<HTMLButtonElement>, index: number, type: 'next' | 'prev') => {
		e.preventDefault();

		return type === 'next' ? setCurrentStep(index + 1) : setCurrentStep(index - 1);
	};

	return (
		<ChakraTabs
			isFitted
			isLazy
			variant='simple'
			index={currentStep}
			onChange={(index) => setCurrentStep(index)}
			position='relative'
			h='full'
		>
			<TabList justifyContent='space-between'>
				{steps.map((step, index) => {
					const { title, number } = step;

					const isCurrentStep = currentStep === index;

					return (
						<Tab key={index} p='15px'>
							<Text
								data-testid={`stepper-step-number-${index + 1}`}
								fontSize='14px'
								color={isCurrentStep ? 'brand.primary' : 'brand.gray600'}
								mr='10px'
							>
								{number}
							</Text>
							<Text
								data-testid={`stepper-step-title-${index + 1}`}
								fontSize='14px'
								color='brand.gray600'
							>
								{title}
							</Text>
						</Tab>
					);
				})}
			</TabList>

			<TabPanels h='calc(100% - 55px)' bgColor='brand.gray100'>
				{steps.map((step, index) => {
					const { children } = step;

					const isFirstStep = index === 0;
					const isLastStep = index === steps.length - 1;

					return (
						<TabPanel
							data-testid={`stepper-step-panel-container-${index + 1}`}
							h='full'
							p='0px'
							key={index}
						>
							<VStack h='full' spacing={4} align='stretch'>
								<Flex flex={1} justify='center' overflowY='auto'>
									{children}
								</Flex>
								<Box h={FOOTER_HEIGHT}>
									<Divider color='brand.gray2' />
									<Flex p='0px 24px' justifyContent='flex-end' h='99%' alignItems='center'>
										<Button
											data-testid={`stepper-step-panel-button-secondary-${index + 1}`}
											variant='secondary'
											mr='20px'
											onClick={
												isFirstStep
													? handleFirstButtonSecondary
													: (e) => handleStep(e, index, 'prev')
											}
										>
											{isFirstStep ? textFirstButtonSecondary : textDefaultButtonSecondary}
										</Button>
										<Button
											data-testid={`stepper-step-panel-button-primary-${index + 1}`}
											disabled={!isValid}
											variant='primary'
											onClick={
												isLastStep ? handleLastButtonPrimary : (e) => handleStep(e, index, 'next')
											}
										>
											{isLastStep ? textLastButtonPrimary : textDefaultButtonPrimary}
										</Button>
									</Flex>
								</Box>
							</VStack>
						</TabPanel>
					);
				})}
			</TabPanels>
		</ChakraTabs>
	);
};

export default Stepper;
