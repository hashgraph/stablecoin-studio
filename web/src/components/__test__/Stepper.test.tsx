import { Box } from '@chakra-ui/react';
import userEvent from '@testing-library/user-event';
import { render } from '../../test/index';
import type { Step } from '../Stepper';
import Stepper from '../Stepper';

const steps: Step[] = [
	{
		number: '01',
		title: 'Option 1',
		complete: false,
		children: <Box>Tab 1</Box>,
	},
	{
		number: '02',
		title: 'Option 2',
		complete: false,
		children: <Box>Tab 2</Box>,
	},
];

const StepsProps = {
	steps,
	textLastButtonPrimary: 'Last button',
	handleLastButtonPrimary: () => jest.fn,
	handleFirstButtonSecondary: () => jest.fn,
	isValid: true,
};

describe(`<${Stepper.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<Stepper {...StepsProps} />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should have number and title each tabs', () => {
		const component = render(<Stepper {...StepsProps} />);

		const titleTab1 = component.getByTestId('stepper-step-title-1');
		const titleTab2 = component.getByTestId('stepper-step-title-2');

		const numberTab1 = component.getByTestId('stepper-step-number-1');
		const numberTab2 = component.getByTestId('stepper-step-number-2');

		expect(titleTab1).toHaveTextContent(steps[0].title);
		expect(titleTab2).toHaveTextContent(steps[1].title);
		expect(numberTab1).toHaveTextContent(steps[0].number);
		expect(numberTab2).toHaveTextContent(steps[1].number);
	});

	test('should change tab when next step button clicked', async () => {
		const component = render(<Stepper {...StepsProps} />);

		const buttonPrimaryTab1 = component.getByTestId('stepper-step-panel-button-primary-1');

		userEvent.click(buttonPrimaryTab1);

		const buttonSecondaryTab2 = component.getByTestId('stepper-step-panel-button-secondary-2');

		expect(buttonSecondaryTab2).toBeInTheDocument();
	});

	test('Cancel button should call function passed by handleFirstButtonSecondary prop', async () => {
		const handleFirstButtonSecondary = jest.fn();

		const component = render(
			<Stepper {...StepsProps} handleFirstButtonSecondary={handleFirstButtonSecondary} />,
		);

		const buttonSecondaryTab1 = component.getByTestId('stepper-step-panel-button-secondary-1');

		userEvent.click(buttonSecondaryTab1);

		expect(handleFirstButtonSecondary).toHaveBeenCalled();
	});

	test('Last button should call function passed by handleLastButtonPrimary prop', async () => {
		const handleLastButtonPrimary = jest.fn();

		const component = render(
			<Stepper {...StepsProps} handleLastButtonPrimary={handleLastButtonPrimary} />,
		);

		const buttonPrimaryTab1 = component.getByTestId('stepper-step-panel-button-primary-1');

		userEvent.click(buttonPrimaryTab1);

		const buttonPrimaryTab2 = component.getByTestId('stepper-step-panel-button-primary-2');

		userEvent.click(buttonPrimaryTab2);

		expect(handleLastButtonPrimary).toHaveBeenCalled();
	});
});
