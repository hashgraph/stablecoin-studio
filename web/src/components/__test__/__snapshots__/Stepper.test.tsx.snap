// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-tabs css-1x6vjw3"
  >
    <div
      aria-orientation="horizontal"
      class="chakra-tabs__tablist css-9jay18"
      role="tablist"
    >
      <button
        aria-controls="tabs-:r0:--tabpanel-0"
        aria-disabled="false"
        aria-selected="true"
        class="chakra-tabs__tab css-h9y905"
        data-index="0"
        id="tabs-:r0:--tab-0"
        role="tab"
        tabindex="0"
        type="button"
      >
        <p
          class="chakra-text css-1ikrtdg"
          data-testid="stepper-step-number-1"
        >
          01
        </p>
        <p
          class="chakra-text css-1i91963"
          data-testid="stepper-step-title-1"
        >
          Option 1
        </p>
      </button>
      <button
        aria-controls="tabs-:r0:--tabpanel-1"
        aria-selected="false"
        class="chakra-tabs__tab css-h9y905"
        data-index="1"
        disabled=""
        id="tabs-:r0:--tab-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1wzl1bo"
          data-testid="stepper-step-number-2"
        >
          02
        </p>
        <p
          class="chakra-text css-1i91963"
          data-testid="stepper-step-title-2"
        >
          Option 2
        </p>
      </button>
    </div>
    <div
      class="chakra-tabs__tab-panels css-27a8z"
    >
      <div
        aria-labelledby="tabs-:r0:--tab-0"
        class="chakra-tabs__tab-panel css-1iyuyap"
        data-testid="stepper-step-panel-container-1"
        id="tabs-:r0:--tabpanel-0"
        role="tabpanel"
        tabindex="0"
      >
        <div
          class="chakra-stack css-1eot3nj"
        >
          <div
            class="css-1c3xk1u"
          >
            <div
              class="css-0"
            >
              Tab 1
            </div>
          </div>
          <div
            class="css-1q6t3f6"
          >
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-12thghs"
            />
            <div
              class="css-1e334ws"
            >
              <button
                class="chakra-button css-hi7zxp"
                data-testid="stepper-step-panel-button-secondary-1"
                type="button"
              >
                Cancel
              </button>
              <button
                class="chakra-button css-qvkb6r"
                data-testid="stepper-step-panel-button-primary-1"
                type="button"
              >
                Next step
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="tabs-:r0:--tab-1"
        class="chakra-tabs__tab-panel css-1iyuyap"
        data-testid="stepper-step-panel-container-2"
        hidden=""
        id="tabs-:r0:--tabpanel-1"
        role="tabpanel"
        tabindex="0"
      />
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
